03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 4, // TODO: check what will happen with: .count = 5
+					.samplers = nullptr // irrelevant for a buffer
+				},
+		};
+
+		nbl::video::IGPUDescriptorSetLayout::SBinding bindingsDS3[1] = {
+				{
+					.binding = 6,
+					.type = nbl::asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 1,
+					.samplers = nullptr // irrelevant for a buffer
+				},
+		};
+		
+		smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout1 = m_device->createDescriptorSetLayout(bindingsDS1);
+		smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout3 = m_device->createDescriptorSetLayout(bindingsDS3);
+		if (!dsLayout1 || !dsLayout3)
+			return logFail("Failed to create a Descriptor Layout!\n");
+
+		// Nabla actually has facilities for SPIR-V Reflection and "guessing" pipeline layouts for a given SPIR-V which we'll cover in a different example
+		smart_refctd_ptr<nbl::video::IGPUPipelineLayout> pplnLayout = m_device->createPipelineLayout({}, nullptr, smart_refctd_ptr(dsLayout1), nullptr,  smart_refctd_ptr(dsLayout3));
+		if (!pplnLayout)
+			return logFail("Failed to create a Pipeline Layout!\n");
+
+		// We use strong typing on the pipelines (Compute, Graphics, Mesh, RT), since there's no reason to polymorphically switch between different pipelines
+		smart_refctd_ptr<nbl::video::IGPUComputePipeline> pipeline;
+		{
+			IGPUComputePipeline::SCreationParams params = {};
+			params.layout = pplnLayout.get();
+			// Theoretically a blob of SPIR-V can contain multiple named entry points and one has to be chosen, in practice most compilers only support outputting one (and glslang used to require it be called "main")
+			params.shader.entryPoint = "main";
+			params.shader.shader = shader.get();
+			// we'll cover the specialization constant API in another example
+			if (!m_device->createComputePipelines(nullptr, { &params,1 }, &pipeline))
+				return logFail("Failed to create pipelines (compile & link shaders)!\n");
+		}
 
-			// Need to know input and output sizes by ourselves obviously
-			constexpr size_t WorkgroupCount = 4096;
-			constexpr size_t BufferSize = sizeof(uint32_t)*WorkgroupSize*WorkgroupCount;
+		// Nabla hardcodes the maximum descriptor set count to 4
+		constexpr uint32_t MaxDescriptorSets = ICPUPipelineLayout::DESCRIPTOR_SET_COUNT;
+		const std::array<IGPUDescriptorSetLayout*, MaxDescriptorSets> dscLayoutPtrs = { nullptr, dsLayout1.get(), nullptr, dsLayout3.get() };
+		std::array<smart_refctd_ptr<IGPUDescriptorSet>, MaxDescriptorSets> ds;
+		auto pool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_NONE, std::span(dscLayoutPtrs.begin(), dscLayoutPtrs.end()));
+		pool->createDescriptorSets(dscLayoutPtrs.size(), dscLayoutPtrs.data(), ds.data());
```

Comment:
should create from pipeline introspection

Reply:
i can only create CPU descriptor set layouts from introspection, is there any way to cast from `ICPUDescriptorSetLayout` to `IGPUDescriptorSetLayout`?

Reply:
no you can't cast, you need to create by hand.

IF you get the example to a mergable state, I can use the asset-converter here maybe.